src/EventStore.Core.Tests/Services/Replication/LeaderReplication/when_replica_subscribes_with_epochs.cs

```diff
@@ -0,0 +1,343 @@
+		private readonly Guid _replicaId = Guid.NewGuid();
+		private TcpConnectionManager _replicaManager;
+		private EpochRecord _lastEpoch;
+		private long _subscribedPosition;
+
+		public override void When() {
+			EpochManager.WriteNewEpoch(0);
+			Writer.Write(CreateLogRecord(0, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(1, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(2, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(3, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(4, "test-stream", "test"), out _);
+			EpochManager.WriteNewEpoch(1);
+
+			_lastEpoch = EpochManager.GetLastEpoch();
+			var epochs = EpochManager.GetLastEpochs(10)
+				.Select(e => new Epoch(e.EpochPosition, e.EpochNumber, e.EpochId)).ToArray();
+
+			_subscribedPosition = _lastEpoch.EpochPosition - 10;
+			AddSubscription(_replicaId, true, epochs, _subscribedPosition, out _replicaManager);
+		}
+
+		[Test]
+		public void subscription_is_sent_a_replica_subscribed_message_from_last_epoch_position() {
+			var subscribed = GetTcpSendsFor(_replicaManager).Select(x => x.Message)
+				.OfType<ReplicationMessage.ReplicaSubscribed>().ToArray();
+			Assert.AreEqual(1, subscribed.Length);
+			Assert.AreEqual(_subscribedPosition, subscribed[0].SubscriptionPosition);
+			Assert.AreEqual(_replicaId, subscribed[0].SubscriptionId);
+			Assert.AreEqual(LeaderId, subscribed[0].LeaderId);
+		}
+	}
+
+	public class when_replica_with_additional_epochs_subscribes_to_position_past_leaders_last_epoch
+		: with_replication_service_and_epoch_manager {
+		private readonly Guid _replicaId = Guid.NewGuid();
+		private TcpConnectionManager _replicaManager;
+		private List<Epoch> _replicaEpochs;
+
+		public override void When() {
+			EpochManager.WriteNewEpoch(0);
+			Writer.Write(CreateLogRecord(0, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(1, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(2, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(3, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(4, "test-stream", "test"), out _);
+			EpochManager.WriteNewEpoch(1);
+
+			_replicaEpochs = new List<Epoch> {
+				new Epoch(1200, 4, Guid.NewGuid()),
+				new Epoch(1100, 3, Guid.NewGuid()),
+				new Epoch(1000, 2, Guid.NewGuid()),
+			};
+			_replicaEpochs.AddRange(EpochManager.GetLastEpochs(10)
+				.Select(e => new Epoch(e.EpochPosition, e.EpochNumber, e.EpochId)).ToList());
+
+			AddSubscription(_replicaId, true, _replicaEpochs.ToArray(), 1200, out _replicaManager);
+		}
+
+		[Test]
+		public void subscription_is_sent_replica_subscribed_message_for_epoch_after_common_epoch() {
+			var subscribed = GetTcpSendsFor(_replicaManager).Select(x => x.Message)
+				.OfType<ReplicationMessage.ReplicaSubscribed>().ToArray();
+			Assert.AreEqual(1, subscribed.Length);
+			Assert.AreEqual(_replicaEpochs[2].EpochPosition, subscribed[0].SubscriptionPosition);
```

Comment:
this one is melting my brain a bit. i think it might only be working because the replica epoch #2 overlaps the common epoch #1 (which wouldn't happen irl)